### ISSUE TITLE
Question about the Two Layer MLP applied to the similarity matrix (i.e `weight_transform` applied to `simi_scores`)

### ISSUE THREAD

**Issue by nikhil-mathew-td** [CVMI-Lab/CoDet#10]
Thanks you for your work! I really appreciate the idea's you've presented in your paper and find them really interesting and useful! I have a question about the paper and was hoping you could provide some clarity. 

In the paper and in the method `_concept_grouping_loss` a two layer MLP is applied to the similarity matrix. It seems to me that for the MLP to learn patterns in the similarity matrix there needs to be some inherent order in the last dimension of the matrix (`mn`). From the paper it seems that mini group is randomly sampled and so there's no specific order maintained in `m` and since `n` depend on the proposals (based on the content of the image) there is no inherent order in `n`.

How does this MLP learn useful features from the similarity matrix? Is there some order maintained during sampling of the mini-group? Is there an ablation study with and without the MLP?

**Comment by machuofan** [comment]
Hi there,

Thanks for your interest in our work. What the MLP does is to predict a weight for each proposal in the query image. For a similarity matrix of dimensions `m*n`, `m` corresponds to the number of proposals in the query image, `n` corresponds to the number of proposals in the support images. Thus, the `i-th` row of the similarity matrix corresponds to similarities between the `i-th` proposal in the query image and all proposals in the support images. Both dimensions (`m` and `n`) are ordered.

Table 4(b) presents the results of corresponding ablation study.

**Comment by nikhil-mathew-td** [comment]
Thanks for your quick response @machuofan . In what way are the dimensions ordered? Are the sampled images sorted in some way for `m` to be ordered or are the proposals sorted in some way for `n` to be ordered?

Table 4(b) does provide an ablation comparing the heuristic approach to the MLP approach. It's clear that this approach leads to better results. I'm just curious as to what operation the MLP is doing to the similarity weight matrix?

**Comment by machuofan** [comment]
The similarity matrix has the same order as the sampled images in dimension `m` and the same order as proposals in dimension `n`. To put it simple, the MLP takes in similarity estimations between one region proposal from the query image and all proposals in the support images, then outputs a value representing the degree of "co-occurrence" of that region proposal.

**Comment by nikhil-mathew-td** [comment]
The issue I'm trying to raise is that if the `m` images are sampled (randomly ordered) and the `n` proposals are sampled from those images (randomly ordered), then information in the `m*n` dimension of  `n X m*n` similarity matrix is randomly ordered. 

As per `codet_fast_rcnn.py`
```python
        self.fc1 = nn.Linear(self.ws_proposal_num * (self.g_size-1), self.ws_proposal_num)
        self.fc2 = nn.Linear(self.ws_proposal_num, 1)
        weight_init.c2_xavier_fill(self.fc1)
        nn.init.normal_(self.fc2.weight, mean=0, std=0.001)
        nn.init.constant_(self.fc2.bias, 0)
        self.weight_transform = nn.Sequential(
            self.fc1, nn.ReLU(inplace=True),
            self.fc2, nn.Flatten(start_dim=0), nn.Softmax(dim=0)
        )
```

The MLP projects this information from `n X m*n` -> `n X n` -> `n X 1` via it's two layers, but if there's no particular order to the `m*n` dimension it should be difficult for the MLP to aggregate in a meaningful way. It must be simply taking an average across the `m*n` and then again across the `n` in the second layer. 

Looking at the weights from the model zoo OV-COCO ResNet50 (see attached [weight_transform_weights.json](https://github.com/CVMI-Lab/CoDet/files/13709359/weight_transform_weights.json)) this seems to be the case the weights are all positive and quite similar (mean of 0.0446 and std_dev of 0.0015 for fc1). The bias for these layers have a bit more variation (mean of -0.0035 and std_dev of 0.0107 for fc1).

It seems like simply averaging across `m*n` and then averaging across `n` then taking the softmax would achieve a similar result. Table 4(b) has a comparison with a handcrafted approach where we take max across `m*n`, then average across `n`, then take an argmax. I was hoping you would have an ablation of simply averaging across both layers.

**Comment by machuofan** [comment]
Yes, I get your point. But an important fact is that, dimension 1 of the similarity matrix is not totally unordered. Here is a simple example: Suppose `n=3`, `m=9`, the number of support images is `m/n=3`. Let `a` and `b` be two rows extracted from the similarity matrix.
```
a = (0.1, 0.1, 0.9, | 0.1, 0.9, 0.1, | 0.1, 0.9, 0.1)
b = (0.9, 0.9, 0.9, | 0.1, 0.1, 0.1, | 0.1, 0.1, 0.1)
```
`a` and `b` have the same mean value, but they imply different extent of co-occurrence. For `a`, the query proposal can find similar proposals in all support images. While for `b`, the query proposal can only find similar proposals in the first support image. Therefore, we would expect `MLP(a) > MLP(b)`. Simply replacing MLP with averaging operation does not work for this case.

If you are curious about the results of averaging operation, you can have a try by replacing the code blocks you mentioned with `torch.mean(similarity_matrix, dim=-1)`.
